src/tty-ask-password-agent/tty-ask-password-agent.c

```diff
@@ -614,8 +696,90 @@ static int parse_argv(int argc, char *argv[]) {
         return 1;
 }
 
+static unsigned int wfa_child(const struct console * con, const unsigned int id)
+{
+        setsid();
+        release_terminal();
+        *usemask |= 1 << id;   /* shared memory area */
+        current_dev = con[id].tty;
+        return id;
+}
+
+static unsigned int wait_for_answer(void)
+{
+        struct console *consoles;
+        struct sigaction sig = {
+                .sa_handler = chld_handler,
+                .sa_flags = SA_NOCLDSTOP | SA_RESTART,
+        };
+        struct sigaction oldsig;
+        sigset_t set, oldset;
+        unsigned int num = 0, id;
+        int status = 0, ret;
+        pid_t job;
+
+        consoles = collect_consoles(&num);
+        if (!consoles) {
+                log_error("Failed to query password: %m");
+                exit(EXIT_FAILURE);
+        }
+        if (num < 2)
+                return wfa_child(consoles, 0);
+
+        assert_se(sigemptyset(&set) == 0);
+        assert_se(sigaddset(&set, SIGHUP) == 0);
+        assert_se(sigaddset(&set, SIGCHLD) == 0);
+        assert_se(sigemptyset(&sig.sa_mask) == 0);
+        assert_se(sigprocmask(SIG_UNBLOCK, &set, &oldset) == 0);
+        assert_se(sigaction(SIGCHLD, &sig, &oldsig) == 0);
+        sig.sa_handler = SIG_DFL;
+        assert_se(sigaction(SIGHUP, &sig, NULL) == 0);
+
+        for (id = 0; id < num; id++) {
+                consoles[id].pid = fork();
+
+                if (consoles[id].pid < 0) {
+                        log_error("Failed to query password: %m");
+                        exit(EXIT_FAILURE);
+                }
+
+                if (consoles[id].pid == 0) {
+                        if (prctl(PR_SET_PDEATHSIG, SIGHUP) < 0)
+                                _exit(EXIT_FAILURE);
+                        zero(sig);
+                        assert_se(sigprocmask(SIG_UNBLOCK, &oldset, NULL) == 0);
+                        assert_se(sigaction(SIGCHLD, &oldsig, NULL) == 0);
+                        return wfa_child(consoles, id);
```

Comment:
If `setsid()` is required, then please mention this in your commit message. Try to explain why you chose to do it the way you did. If this is really the case, then yes, a fork() sounds like a reasonable solution.

Anyway, if it is really necessary to make the terminal the controlling TTY, then yes, please use fork(). However, I really dislike the coding style. Please fork one process per TTY, which has a proper function that performs whatever the child does and then exits. Don't return to the caller after fork().

Then make the parent wait for the first child and abort the other children.